backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt

```diff
@@ -901,31 +901,48 @@ internal class FunctionGenerationContext(val function: LLVMValueRef,
             }
         }
 
-        // If object is imported - access it via getter function.
-        if (isExternal(irClass)) {
-            val valueGetterName = irClass.objectInstanceGetterSymbolName
-            val valueGetterFunction = LLVMGetNamedFunction(context.llvmModule, valueGetterName) ?:
-                LLVMAddFunction(context.llvmModule, valueGetterName,
-                        functionType(kObjHeaderPtr, false, kObjHeaderPtrPtr))
-            return call(valueGetterFunction!!,
-                        listOf(),
-                        resultLifetime = Lifetime.GLOBAL,
-                        exceptionHandler = exceptionHandler)
+        val storageKind = irClass.storageKind(context)
+
+        val objectPtr = if (isExternal(irClass)) {
+            when (storageKind) {
+                // If thread local object is imported - access it via getter function.
+                ObjectStorageKind.THREAD_LOCAL -> {
+                    val valueGetterName = irClass.threadLocalObjectInstanceGetterSymbolName
+                    val valueGetterFunction = LLVMGetNamedFunction(context.llvmModule, valueGetterName)
+                            ?: LLVMAddFunction(context.llvmModule, valueGetterName,
+                                    functionType(kObjHeaderPtrPtr, false))
+                    call(valueGetterFunction!!,
+                            listOf(),
+                            resultLifetime = Lifetime.GLOBAL,
+                            exceptionHandler = exceptionHandler)
+                }
+
+                // If global object is imported - import it's storage directly.
+                ObjectStorageKind.PERMANENT, ObjectStorageKind.SHARED -> {
+                    val llvmType = getLLVMType(irClass.defaultType)
+                    importGlobal(
+                            irClass.globalObjectInstanceSymbolName,
+                            llvmType,
+                            origin = irClass.llvmSymbolOrigin
+                    )
+                }
+            }
+        } else {
+            // Local globals and thread locals storage info is stored in our map.
+            val singleton = context.llvmDeclarations.forSingleton(irClass)
+            val instanceAddress = singleton.instanceStorage
+            instanceAddress.getAddress(this)
         }
 
-        val storageKind = irClass.storageKind(context)
         when (storageKind) {
-            ObjectStorageKind.SHARED -> context.llvm.sharedObjects += irClass
+            ObjectStorageKind.SHARED -> context.llvm.sharedObjects += objectPtr
             ObjectStorageKind.THREAD_LOCAL -> context.llvm.objects += irClass
```

Comment:
We need to recheck if `objects` set is still required.
Motivation: the usage before this PR was kinda a hack to decide on (de)initializers generation, this hack probably don't work as expected anymore.